sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -239,18 +275,47 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
 
     def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
 
-      // --- BroadcastHashJoin --------------------------------------------------------------------
+      // --- Hints specified, choose join strategy based on hints. --------------------------------
 
-      // broadcast hints were specified
+      // broadcast hints specified with equi-join keys, use broadcast-hash
       case ExtractEquiJoinKeys(joinType, leftKeys, rightKeys, condition, left, right, hint)
-        if canBroadcastByHints(joinType, left, right, hint) =>
+          if canBroadcastByHints(joinType, left, right, hint) =>
         val buildSide = broadcastSideByHints(joinType, left, right, hint)
         Seq(joins.BroadcastHashJoinExec(
           leftKeys, rightKeys, joinType, buildSide, condition, planLater(left), planLater(right)))
 
-      // broadcast hints were not specified, so need to infer it from size and configuration.
+      // broadcast hints specified with no equi-join keys, use broadcast-nested-loop
+      case j @ logical.Join(left, right, joinType, condition, hint)
+          if canBroadcastByHints(joinType, left, right, hint) =>
+        val buildSide = broadcastSideByHints(joinType, left, right, hint)
+        Seq(joins.BroadcastNestedLoopJoinExec(
+          planLater(left), planLater(right), buildSide, joinType, condition))
+
+      // shuffle-merge hints specified
+      case ExtractEquiJoinKeys(joinType, leftKeys, rightKeys, condition, left, right, hint)
+          if canShuffleMergeByHints(leftKeys, hint) =>
+        Seq(joins.SortMergeJoinExec(
+          leftKeys, rightKeys, joinType, condition, planLater(left), planLater(right)))
+
+      // shuffle-hash hints specified
+      case ExtractEquiJoinKeys(joinType, leftKeys, rightKeys, condition, left, right, hint)
+          if canShuffleHashByHints(joinType, left, right, hint) =>
+        val buildSide = shuffleHashSideByHints(joinType, left, right, hint)
+        Seq(joins.ShuffledHashJoinExec(
+          leftKeys, rightKeys, joinType, buildSide, condition, planLater(left), planLater(right)))
+
+      // shuffle-replicate-nl hints specified
+      case logical.Join(left, right, _: InnerLike, condition, hint)
+          if shuffleReplicateNLByHints(hint) =>
+        Seq(joins.CartesianProductExec(planLater(left), planLater(right), condition))
+
+
+      // --- No hints specified, choose join strategy based on size and configuration. ------------
```

Comment:
The `JoinSelection` class is getting large, so how about moving this class into a separate one?

Reply:
I'm good either way. The argument for not moving this would be Join is one of the basic operators.

Reply:
It will lose the change history. Thus, I would keep it in this file.